[aws-datastore] Implement DataStore clear feature

rjuliano opened the pull request:
*Issue #, if available:*

*Description of changes:*

# DataStore clear
Adding clear method to the DataStore plugin.

The intent of the functionality of the clear method is to blow away the entire underlying data store and reset it as needed. The current implementation of the DataStore plugin uses SQLite for the implementation of the LocalStorageAdapter interface.

## Summary of changes

* Added `clear` method to the `DataStoreCategory`, `DataStoreCategoryBehavior`, and `LocalStorageAdapter`
* Implement `clear` for `AWSDataStorePlugin`. It stops the orchestrator, invokes the clear method of the storage adapter, ~~and restarts the orchestrator~~.  The orchestrator is not restarted right away; instead, it is restarted upon the next interaction with the DataStore category.
* Implement `clear` for `SQLiteStorageAdapter`. 
  * The implementation closes the adapter and the current handle to the DB connection. Then it deletes the database file using the context and attempts to re-initialize the adapter.
   * Keep a reference to the `Context` object received in the `configure` method so we can use it to get access to the database file for clearing.

* Updated some of the test classes that are used to create a synchronous interface to the storage adapter.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jamesonwilliams commented:
I think if you `git rebase --pull origin master` or `git merge origin/master`, the integ tests should be ~passing now. @raphkim [\@Ignore'd the broken REST API tests, here](https://github.com/aws-amplify/amplify-android/pull/411/files).

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/SynchronousStorageAdapter.java:
```diff
@@ -285,4 +286,14 @@ public void terminate() throws DataStoreException {
+     * either completes or throws an exception.
+     */
+    public void clear() {
+        Completable.fromObservable(emitter -> {
```
Ah yea, good call. The `Await.result(...)` thing could basically be replaced with a `Single` and a `blockingGet()`. At the time I wrote `Await.result(...)` we hadn't started using `Rx` from the tests across different modules. Here in the DataStore tests, its doubly okay, though, since the actual _source code_ uses Rx.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
2020

wooj2 replied:
nit: date change from. 2019 to 2020

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class SQLiteStorageAdapterClearTest {
```
`final`

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+    @Before
+    public void setup() {
+        TestStorageAdapter.cleanup();
+        context = (ApplicationProvider.getApplicationContext());
```
Drop the outer `(`,`)`

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        context = (ApplicationProvider.getApplicationContext());
+        adapter = TestStorageAdapter.create(AmplifyModelProvider.getInstance());
+        observer = TestObserver.create();
+        adapter.observe().subscribe(observer);
```
There's a short-hand for this, [`Observable#test()`](http://reactivex.io/RxJava/javadoc/rx/Observable.html#test--):

```
TestObserver<StorageItemChange.Record> observer = adapter.observe().test();
```

It only saves a line there, but it doesn't need to be disposed, which is neat. And it's _slightly_ more expressive of the intention.

rjuliano replied:
Updating this one. I had made the change locally then stashed it and ended up getting lost.

wooj2 commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:
```diff
@@ -390,6 +390,22 @@ public void observe(
+     * Stops all synchronization processes and deletes the
+     * SQLite database from device's local storage.
+     * @param onComplete Invoked if the call is successful.
+     * @param onError Invoked if an exception occurs.
```
Seems like this should be changed to: "Invoked if not successful" ?

rjuliano replied:
Yeah...that makes more sense. Updating it.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        observer.assertValueCount(1);
+
+        adapter.clear();
+        assertFalse(context.getDatabasePath(DATABASE_NAME).exists());
```
Nice this reads really well

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        observer.assertNotSubscribed();
+        //Assert the observer was terminated during the file removal.
+        observer.assertTerminated();
+    }
```
What happens if you do `adapter.query(BlogOwner.class)` at this point?

rjuliano replied:
Prior to my latest commits, it would just blow up. I've added code to more gracefully reset the local storage adapter after the file is removed.

wooj2 commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/LocalStorageAdapter.java:
```diff
@@ -186,4 +186,14 @@ Cancelable observe(
+
+    /**
+     * Each implementation of this adapter interface will have its own
+     * interpretation of what clear means. At a high level, the intent is to destroy
```
1.  Remove "At a high level"
2.  In iOS, we are not making any guarantees on in-memory objects...  We still have some objects still in memory (that have read in configuration) so that customers  do not need to call configure() on amplify again if they want to use Datastore.
3.  This comment seems to be duplicated from the "*behavior.java" class... I'm not familiar with your code base, but my preference would be to only have it at the interface level.

rjuliano replied:
I've updated this. We have a little bit of duplication because we have an interface for the plugin as well as an interface for the storage adapter.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        observer.assertNotSubscribed();
+        assertFalse(context.getDatabasePath(DATABASE_NAME).exists());
+
+        adapter.clear();
```
Ah, this is the actual `// Act` part of the test, right? That `clear()` can be called multiple times without raising hell? I missed it until my second read - can you highlight this via a comment or something?

rjuliano replied:
I modified the test quite a bit in the latest couple of commits. Added a few comments on the assertions.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:
```diff
@@ -390,6 +390,22 @@ public void observe(
 
+    /**
+     * Stops all synchronization processes and deletes the
+     * SQLite database from device's local storage.
```
Maybe a note about how/when this stuff is able to resume, afterwards? What if I have some changes in the outbound mutations queue? Will they be sent to AppSync, still?

rjuliano replied:
Updated the comment and made a note to come up with a test to verify this.

wooj2 commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+
+import io.reactivex.observers.TestObserver;
+
+import static com.amplifyframework.datastore.storage.sqlite.SQLiteStorageAdapter.DATABASE_NAME;
```
For unit tests, I think it is a better test not to share constants between test code and production code.

jamesonwilliams replied:
^ I do agree. I made this change in my own tests probably 1-2 years ago.

It's a tradeoff of maintaining two copies of a constant, versus keeping the constant as an implementation detail (`private` field access, etc.) 

If the source code version changes, a self-maintained version in test would be able to catch it. But, if you just use the constant, you could have like:

```
@VisibleForTesting static final String GOOD_VALUE = "oopsActuallyBadPsych!";
```

If test still had a hardcoded `"goodValue"` somewhere, it'd catch it.

rjuliano replied:
Fixing this.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/LocalStorageAdapter.java:
```diff
@@ -186,4 +186,14 @@ Cancelable observe(
     void terminate() throws DataStoreException;
+
+    /**
+     * Each implementation of this adapter interface will have its own
```
👍 Good docs.

jpignata replied:
Agree! These are good docs!

wooj2 commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        //check that it's not subscribed anymore.
+        observer.assertNotSubscribed();
+        //Assert the observer was terminated during the file removal.
+        observer.assertTerminated();
```
is there a way to assert that the success handler gets called?

rjuliano replied:
We created a [wrapper class](https://github.com/aws-amplify/amplify-android/blob/782eb601676c1e94b85718c4a66028784e90a802/aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/SynchronousStorageAdapter.java#L296) for that we use for testing that basically blocks until either the success handler gets called or bubbles up an exception. So if the call does not throw, we know the success handler was called.

wooj2 commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,111 @@
+        observer.assertNotSubscribed();
+        assertFalse(context.getDatabasePath(DATABASE_NAME).exists());
+
+        adapter.clear();
```
is there a way to make sure that the error handler gets called?

rjuliano replied:
See comment above. The adapter is wrapped by a class that blocks the call until it either succeeds or throws.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -579,6 +579,25 @@ public synchronized void terminate() throws DataStoreException {
+            onComplete.call();
+        } catch (Exception exception) {
+            DataStoreException dataStoreException = new DataStoreException(
+                "Error while trying to delete database from local storage.",
```
Maybe, for more clarity, `"Error while trying to clear data from the local DataStore storage"` ?

rjuliano replied:
Updated message.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -579,6 +579,25 @@ public synchronized void terminate() throws DataStoreException {
+    @Override
+    public void clear(@NonNull Action onComplete,
+                      @NonNull Consumer<DataStoreException> onError) {
+        try {
+            terminate();
+            sqliteStorageHelper.deleteDatabaseFromDisk();
+            onComplete.call();
+        } catch (Exception exception) {
```
Let's try to make this exception handling more narrow. Which of these calls are actually expected to throw, and what types of exceptions will they throw?

Can you do:
```
try {
    justTheThingsThatThrow();
} catch (OnePossibleCaseException | AnotherCaseException failureToClear) {
    // wrap in DataStoreException and re-throw
}
```

rjuliano replied:
This method has been changes. Take a look at the updated logic and let me know.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageHelper.java:
```diff
@@ -172,6 +175,13 @@ public synchronized void update(
+     * Deletes the SQLite database file from the device's local storage.
+     */
+    public void deleteDatabaseFromDisk() {
+        context.deleteDatabase(getDatabaseName());
```
Should you call `super.close()` first, to be safe?

rjuliano replied:
I actually removed this code. I'm calling the helper's close method from the adapter where I can have better control to restart things

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+        assertTrue(context.getDatabasePath(DATABASE_NAME).exists());
+    }
+
+    private static final class TestFileObserver extends FileObserver {
```
Ah, cool beans. I've never seen or used the [`FileObserver`](https://developer.android.com/reference/android/os/FileObserver), before! Good stuff.

rjuliano replied:
Yeah...I initially used a method that would return me the created date of the file as an instance of `FileTime`, but it appears creation date was not available as an attribute in older versions of Android.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+        assertDbFileExists();
+        assertEquals(0, fileObserver.createFileEventCount);
+        assertEquals(0, fileObserver.deleteFileEventCount);
+        BlogOwner blogger1 = createBlogger("Dummy Blogger Sr.");
+        BlogOwner blogger2 = createBlogger("Dummy Blogger Jr.");
```
:-D

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+        //Verify observer is still alive
+        assertFalse(subscriberDisposableRef.get().isDisposed());
+        assertObserverReceivedRecord(blogger1);
+        //The ensures files are created at least 1 second apart
+        Thread.sleep(1000);
```
But how come? Is it necessary?

rjuliano replied:
It's not needed anymore. Before I started using the `FileObserver`, I was checking the created date of the file. Had to move away from that because creation date attribute is not available in older versions of Android.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+        BlogOwner blogger2 = createBlogger("Dummy Blogger Jr.");
+        //Save a record and check if it's there
+        adapter.save(blogger1);
+        assertRecordIsInDb(blogger1.getName());
```
In theory you could achieve this with:
```
assertEquals(blogger1, adapter.query(BlogOwner.class, blogger1.getId()).get(0));
```
But that gets kind of long, and loses the DAMP you have.

Maybe:
```
assertRecordIsInDB(BlogOwner.class, blogger1.getId());
```
To use the same type of arguments as the storage interface usually do, though?
```
private <T extends Model> void assertRecordIsInDB(T item) {
    List<T> results = adapter.query(item.getClass(), item.getId());
    assertEquals(1, results.size());
    assertEquals(item, results.get(0));
}
```

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+        adapter.terminate();
+        //Verify observer was disposed.
+        assertTrue(subscriberDisposableRef.get().isDisposed());
+
```
nit: an extra empty

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,210 @@
+
+    private void assertObserverReceivedRecord(BlogOwner blogger) {
+        for (StorageItemChange<? extends Model> owner : observer.values()) {
+            if (owner.item() instanceof BlogOwner &&
```
There is also an `item.itemClass()`, you could do like `BlogOwner.class.isAssignableFrom(item.itemClass())` or something like that. (Or, `BlogOwner.class.equals(item.itemClass())`.)

rjuliano replied:
I'm going with `BlogOwner.class.isAssignableFrom(item.itemClass())`.

rjuliano commented:
> Looks pretty good to me. Have a question about the re-init part, inline. Love the main test case you wrote out!

I made a change to the logic that controls the re-initialization of the synchronization processes. It's no longer restarting right away. Instead, it restarts on the next interaction with the DataStore.

I believe I addressed most of the comments, but let me know if I missed anything. In the process of changing the way we restart the sync process, i realized there were some gaps and no real test to verify. 

The most significant change of this batch of changes is in `AWSDataStorePluginTest`. I added a test to verify the restart behavior.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -568,6 +572,35 @@ public synchronized void terminate() throws DataStoreException {
+
+        try {
+            context.deleteDatabase(DATABASE_NAME);
+        } catch (Exception exception) {
```
Can we tighted this up at all? `SQLSpecialException | SQLOtherSpecialException sqlError)` ?

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -95,22 +98,38 @@ public Orchestrator(
+    public synchronized Completable start() {
+        return Completable.fromSingle(single -> {
+            try {
+                if (!storageObserver.isObserving()) {
```
Let's align the naming of all of these to their corresponding `start<whatever>` methods. Here, this would be come:
```
storageObserver.isObservingStorageChanges()
```

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -95,22 +98,38 @@ public Orchestrator(
+                    LOG.debug("Starting local storage observer.");
+                    storageObserver.startObservingStorageChanges();
+                }
+                if (!subscriptionProcessor.isObserving()) {
```
And this one,
```
subscriptionsProcessor.isSubcribed()
```
Or
```
subscriptionsProcessor.hasSubscriptions()
```

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -95,22 +98,38 @@ public Orchestrator(
+                    subscriptionProcessor.startSubscriptions();
+                }
+                syncProcessor.hydrate().blockingAwait();
+                if (!mutationProcessor.isObserving()) {
```
```
mutationProcessor.isDrainingMutationOutbox()
```

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -95,22 +98,38 @@ public Orchestrator(
+                    LOG.debug("Starting mutation processor.");
+                    mutationProcessor.startDrainingMutationOutbox();
+                }
+                subscriptionProcessor.startDrainingMutationBuffer();
```
How come this one doesn't need an `if` guard like the others?

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -95,22 +98,38 @@ public Orchestrator(
-            mutationProcessor.startDrainingMutationOutbox();
-            subscriptionProcessor.startDrainingMutationBuffer();
+    public synchronized Completable start() {
+        return Completable.fromSingle(single -> {
```
Would this be better as just:
```
Completable.create(emitter -> {
    try {
        tryToSucceed();
        emitter.onComplete();
    } catch (WhateverException whatever) {
       emitter.onError(whatever);
    } 
});
```

`fromCallable` or `fromAction` or `forRunnable` ... one of these factories must do the exception handling for you, I think, too?

like,
```
Completable.fromAction(() -> {
}).doOnError(thrownInsideOfAction -> {
    // I need to look up if it works like this ...
});
```

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java:
```diff
@@ -84,6 +84,15 @@ void startDrainingMutationOutbox() {
         );
     }
 
+    /**
+     * Checks if the mutation processor is listening
+     * for events.
+     * @return True if there are listeners. False otherwise.
+     */
+    boolean isObserving() {
+        return disposable.size() > 0;
+    }
```
```suggestion
    /**
     * Checks if the mutation processor is actively observing the mutation outbox.
     * @return True if the mutation processor is subscribed the mutation outbox.
     */
    boolean isDrainingMutationOutbox() {
        return !disposable.isEmpty();
    }
```

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessor.java:
```diff
@@ -123,6 +125,18 @@ void startSubscriptions() {
         .onErrorResumeNext(next -> {
             next.onComplete();
         })
+        .doOnDispose(() -> {
+            // When the observable is disposed, we need to call cancel() on the subscription
+            // so it can properly dispose of resources if necessary. For the AWS API plugin,
+            // this means means closing the underlying network connection.
+            synchronized (cancelable) {
+                Cancelable subscriptionOperation = cancelable.get();
+                if (subscriptionOperation != null) {
+                    LOG.info("Terminating subscription operation.");
+                    subscriptionOperation.cancel();
+                }
+            }
+        })
```
L108 was _supposed_ to do this.

```
            emitter.setCancellable(cancelable::get);
```

However, I think I found elsewhere, that `setCancellable(...)` doesn't do what I thought it was supposed to do. There is also `emitter.setDispoable(Disposable...)`, though. I have a factory method elsewhere in the code base that builds a `Disposable` from our own `Cancelable` type. There's a note about "pulling it out into a utility class". Maybe its time.....

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessor.java:
```diff
@@ -149,6 +163,15 @@ void startDrainingMutationBuffer() {
         );
     }
 
+    /**
+     * Checks if the subscription processor is listening
+     * for events.
+     * @return true if there are listeners. False otherwise.
+     */
+    boolean isObserving() {
+        return disposable.size() > 0;
+    }
+
```
This one is tricker than in the other case, because the `SubscriptionProcessor` has an intermediate state, where it is buffering content, but is not processing it. So, we should decided which state we mean. From your current implementation, it would match two states:
- subscribed, not processing
- subscribed, processing under way

```suggestion
    /**
     * Checks if the subscription processor is currently buffering events from the remote AppSync instance.
     * @return true if the subscription processor has active subscriptions to AppSync
     */
    boolean isSubscribed() {
        return !disposable.isEmpty();
    }

```
The other meaning would also be easy to implement:
```
/**
 * Checks if the subscription processor is actively processing events found on its AppSync subscriptions.
 * @return true if the subscription processor is actively processing subscription events
 */
public isDrainingEventBuffer() {
    return disposable.size() == 2;
}
```

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -74,16 +84,17 @@
-            .addModel(Person.class)
-            .build());
+        modelProvider = spy(AmplifyCliGeneratedModelProvider.singletonInstance());
+        subscriptionCancelledCounter = new AtomicInteger(0);
+        subscriptionStartedCounter = new AtomicInteger(0);
```
0 is already the default for [`AtomicInteger()`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html#AtomicInteger--)

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -158,29 +168,117 @@ public void configureAndInitializeInApiModeWithoutApi() throws JSONException, Am
+                    single.onSuccess(true);
+                }, single::onError);
+            })
+        ).blockingGet(OPERATION_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Does this construct work? It looks like you're making `assertFoo` statements on a different thread, which the jUnit test runner thread would not capture.

 If you put a `fail()` inside one these things, does the test runner thread catch it?

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -158,29 +168,117 @@ public void configureAndInitializeInApiModeWithoutApi() throws JSONException, Am
+            awsDataStorePlugin.save(person1, itemSaved -> {
+                assertNotNull(itemSaved.item().getId());
+                assertEquals(person1.getLastName(), itemSaved.item().getLastName());
+                single.onSuccess(true);
```
Since you ignore the result of the single, can't you just use `Completable.create`, `Completable.fromAction`, `Completable.fromRunnable`, etc?

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -158,29 +168,117 @@ public void configureAndInitializeInApiModeWithoutApi() throws JSONException, Am
 
-        assertTrue(syncProcessorInvoked);
+    private void assertRemoteSubscriptionsStarted() {
+        // For each model, there should be 3 subscriptions setup.
```
You could phrase this as,
```
final int diffTypesOfSubscriptions  = SubscriptionType.values().length;
```
and then get out of hte business of writing `3`, while also clarifying its meaning.

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -194,16 +292,26 @@ private static ApiCategory mockApiCategoryWithGraphQlApi() throws AmplifyExcepti
 
         // Make believe that subscriptions return response immediately
         doAnswer(invocation -> {
-            int indexOfStartConsumer = 2;
```
Oh dang, looks like this wasn't doin' much before huh lol

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -194,16 +292,26 @@ private static ApiCategory mockApiCategoryWithGraphQlApi() throws AmplifyExcepti
+            int indexOfStartConsumer = 1;
+            Consumer<String> onStart = invocation.getArgument(indexOfStartConsumer);
+            GraphQLOperation<?> mockOperation = mock(GraphQLOperation.class);
+            doAnswer(opAnswer -> {
+                this.subscriptionCancelledCounter.incrementAndGet();
+                return null;
+            }).when(mockOperation).cancel();
```
Oh fuck, a `doAnswer` inside of a `doAnswer`, Head just exploded.

The situation with the `Cancelable` return type and `String` "on started" token is such a nightmare, isn't it? On a brighter day, maybe we could cleanup that API signature so we don't have to get so clever about managing/mocking/testing it.

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -568,6 +573,40 @@ public synchronized void terminate() throws DataStoreException {
+                                   @NonNull Consumer<DataStoreException> onError) {
+        try {
+            LOG.debug("Shutting down thread pool for the storage adapter.");
+            threadPool.awaitTermination(THREAD_POOL_TERMINATE_TIMEOUT, TimeUnit.MILLISECONDS);
```
Don't we need to explicitly shut down the threadpool?

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -94,32 +97,88 @@ public Orchestrator(
-                LOG.info("Cloud synchronization is now fully active.");
-            }));
+        // Only start if it's stopped.
+        if (OrchestratorStatus.STOPPED.equals(status.get())) {
```
I really like the use of finite state machines in the iOS codebase.. we should nick some of that.

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -568,6 +573,40 @@ public synchronized void terminate() throws DataStoreException {
+            threadPool.awaitTermination(THREAD_POOL_TERMINATE_TIMEOUT, TimeUnit.MILLISECONDS);
+            LOG.debug("Storage adapter thread pool shutdown.");
+        } catch (InterruptedException exception) {
+            LOG.warn("Storage adapter thread pool was interrupted during shutdown.", exception);
```
Was it interrupted or just didn't stop within some timeframe?

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -568,6 +573,40 @@ public synchronized void terminate() throws DataStoreException {
+        initialize(context, schemaList -> {
+            onComplete.call();
+        }, exception -> {
+                onError.accept(new DataStoreException("Error occurred whilte trying to re-initialize " +
```
maybe move the string to a new line instead of concatenating?

rjuliano replied:
Done.

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -94,32 +97,88 @@ public Orchestrator(
-        storageObserver.stopObservingStorageChanges();
-        subscriptionProcessor.stopAllSubscriptionActivity();
-        mutationProcessor.stopDrainingMutationOutbox();
+        if (OrchestratorStatus.STARTED.equals(status.get())) {
```
Can you use the convenience method you created with this body above?

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AsyncUtils.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public final class AsyncUtils {
+    private AsyncUtils() {
+
```
Nit: extra line?

jpignata commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java:
```diff
@@ -185,12 +189,16 @@ public void configure(
+            .andThen(initializeOrchestrator())
+            .blockingGet(PLUGIN_INIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        if (initError != null) {
+            throw new AmplifyException("Failed to initialize the local storage " +
```
Nit: maybe newline instead of concat?

jpignata replied:
Why is this more likely to happen after clear?

jamesonwilliams replied:
@jpignata I don't think its more likely. I have this code in one of my PRs, too, though. Probably, we could carve this out into its own targeted PR. (Not saying we should, at this point, just that it would be possible.)

@rjuliano and I have been talking offline about reducing/removing all of our dangerous `blockPermanently()` type calls, which can hose threads. We only have a few such calls in source code, but there are quite a few in the tests, right now. We should start whittling away at them by providing some sane timeouts, so that nothing is wedged with permanence.

jamesonwilliams commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,217 @@
+
+        /**
+         * The event handler, which must be implemented by subclasses.
+         *
```
I believe the best practice is _not_ to provide Javadocs for `@Override` methods, since it will be mostly duplicative of the content of the interface/base class being overriden.

The [`JavadocMethod`](https://checkstyle.sourceforge.io/config_javadoc.html#JavadocMethod) checkstyle has this a default setting:

> Javadoc is not required on a method that is tagged with the @Override annotation.

In a few place in this code base, we have the `@inheritDoc` Javadoc directive on `@Override`n methods.

A guy on [StackOverflow](https://stackoverflow.com/a/3062697/695787) goes on:

> I studied this problem during my PhD and found that in general folks will never be aware of the extra information in the overriding version if they are invoking through a supertype.

Holy Cow. Talk about "Diving Deep!"

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/AsyncUtils.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Utility class for functions used throughout DataStore.
+ */
+public final class AsyncUtils {
```
I've been working hard for us to avoid having any `FooUtils` classes (by name).

Would it make sense to call this class `Cancelables`? (with one `l`, like our `Cancelable`, and unlike the Java concurrency `Cancellable`.)

It would read like:

```
Disposable disposables = Cancelables.asDisposable(cancelable);
```

The advantage here is that `Cancelables` is only ever going to have a few things in it, and they'll all relate to that data type.

`AsyncUtils` could end up being abused in the future, and have all kinds of executors, schedulers, etc., in it., over time.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapter.java:
```diff
@@ -568,6 +573,43 @@ public synchronized void terminate() throws DataStoreException {
+                                   @NonNull Consumer<DataStoreException> onError) {
+        try {
+            LOG.debug("Shutting down thread pool for the storage adapter.");
+            threadPool.shutdown();
+            if (!threadPool.awaitTermination(THREAD_POOL_TERMINATE_TIMEOUT, TimeUnit.MILLISECONDS)) {
+                threadPool.shutdownNow();
+            }
```
This solved a bug, right? This must have been very tricky to find. Nice work.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -94,32 +97,88 @@ public Orchestrator(
-                LOG.info("Cloud synchronization is now fully active.");
-            }));
+        // Only start if it's stopped.
+        if (OrchestratorStatus.STOPPED.equals(status.get())) {
```
To help with the indentation / nesting, can we invert this?

```
if (!OrchestratorStatus.STOPPED.equals(status.get())) {
    return initializationCompletable;
}
// Save 4 cols, here ...
```

You can save another eight more cols with the `andThen(` on L120, too, by:
```
initializationCompletable = mutationOutbox.load().andThen(Completable.fromAction(() -> {
    // 8 left of where it was 
}
```

rjuliano replied:
Heck yeah...that's much better.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java:
```diff
@@ -94,32 +97,88 @@ public Orchestrator(
+     */
+    enum OrchestratorStatus {
+        /**
+         * The orchestrator is stopping.
```
True 'nuff, but when it is stopping, what does that mean for the system -- request to `start()` will be denied? What are the valid input and output states, from this step?

Same Q for all

rjuliano replied:
I went through and updated the comments for each of those to be a bit more descriptive.

jamesonwilliams commented on aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SubscriptionProcessor.java:
```diff
@@ -114,15 +113,19 @@ void startSubscriptions() {
                 emitter::onComplete
-            ));
+            );
+            // When the observable is disposed, we need to call cancel() on the subscription
+            // so it can properly dispose of resources if necessary. For the AWS API plugin,
+            // this means means closing the underlying network connection.
+            emitter.setDisposable(
+                AsyncUtils.asDisposable(subscriptionCancelable)
+            );
```
I think this probably solves a bug I've been seeing lately. Thanks!

rjuliano replied:
Yeah...it's a pesky one too.

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -21,49 +21,60 @@
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
+import static org.mockito.ArgumentMatchers.argThat;
```
Can you use the one on the `org.mockito.Mockito` facade, instead? `org.mockito.Mockito.argThat`?

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -21,49 +21,60 @@
 public final class AWSDataStorePluginTest {
-    private static final long OPERATION_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(1);
+    private static final String MOCK_API_PLUGIN_NAME = "MockApiPlugin";
+    private static final int ASSERTION_TIMEOUT_MS = 1000;
```
` = 1_000 /* ms */`, or ` = TimeUnit.SECONDS.toMillis(1);`

jamesonwilliams commented on aws-datastore/src/test/java/com/amplifyframework/datastore/AWSDataStorePluginTest.java:
```diff
@@ -127,60 +137,144 @@ public void configureAndInitializeInApiModeWithoutApi() throws JSONException, Am
+        Person result2 = synchronousDataStore.get(Person.class, person2.getId());
+        assertEquals(person2, result2);
+
+        verify(mockApiCategory, timeout(ASSERTION_TIMEOUT_MS).atLeastOnce())
```
Oooh! I've never see this before. Nice find: `timeout(value).atLeastOnce()`

rjuliano commented on core/src/main/java/com/amplifyframework/datastore/DataStoreCategoryBehavior.java:
```diff
@@ -225,4 +225,14 @@ void observe(
     );
+
+    /**
+     * Every class that implements this behavior will have its own
```
Updated the docs.

rjuliano commented on aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterClearTest.java:
```diff
@@ -0,0 +1,217 @@
+
+        @SuppressWarnings("deprecation")
+        TestFileObserver(@NonNull String path) {
+            super(path, FileObserver.CREATE | FileObserver.DELETE | FileObserver.DELETE_SELF);
```
I wasn't sure what Android's `deleteDatabase` method would do, so I figured I'd look for either one. It deletes both the database and the journal file as well.
